src/index.tsx

```diff
@@ -28,29 +33,164 @@ export interface Options {
   excludePaths?: string[];
 }
 
-export interface MatchOptions {
-  exact?: boolean;
-  strict?: boolean;
-  sensitive?: boolean;
+export interface BreadcrumbMatch<ParamKey extends string = string> {
+  /**
+   * The names and values of dynamic parameters in the URL.
+   */
+  params: Params<ParamKey>;
+  /**
+   * The portion of the URL pathname that was matched.
+   */
+  pathname: string;
+  /**
+   * The pattern that was used to match.
+   */
+  pattern: PathPattern;
+  /**
+   * The route object that was used to match.
+   */
+  route?: BreadcrumbsRoute;
 }
 
-export interface BreadcrumbsRoute {
-  path: string;
-  breadcrumb?: React.ComponentType | React.ElementType | string | null;
-  matchOptions?: MatchOptions;
-  routes?: BreadcrumbsRoute[];
+export interface BreadcrumbComponentProps<ParamKey extends string = string> {
+  key: string;
+  match: BreadcrumbMatch<ParamKey>;
+  location: Location;
+}
+
+export type BreadcrumbComponentType<ParamKey extends string = string> =
+  React.ComponentType<BreadcrumbComponentProps<ParamKey>>;
+
+export interface BreadcrumbsRoute<ParamKey extends string = string>
+  extends RouteObject {
+  children?: BreadcrumbsRoute[];
+  breadcrumb?: BreadcrumbComponentType<ParamKey> | string | null;
   props?: { [x: string]: unknown };
 }
 
-export interface BreadcrumbData {
-  match: { url: string };
+export interface BreadcrumbData<ParamKey extends string = string> {
+  match: BreadcrumbMatch<ParamKey>;
   location: Location;
   key: string;
   breadcrumb: React.ReactNode;
 }
 
-const DEFAULT_MATCH_OPTIONS = { exact: true };
-const NO_BREADCRUMB = 'NO_BREADCRUMB';
+// The code below is modified from React Router
+interface BreadcrumbsRouteMeta {
+  relativePath: string;
+  childrenIndex: number;
+  route: BreadcrumbsRoute;
+}
+
+interface BreadcrumbsRouteBranch {
+  path: string;
+  score: number;
+  routesMeta: BreadcrumbsRouteMeta[];
+}
+
+const joinPaths = (paths: string[]): string => paths.join('/').replace(/\/\/+/g, '/');
+
+const paramRe = /^:\w+$/;
+const dynamicSegmentValue = 3;
+const indexRouteValue = 2;
+const emptySegmentValue = 1;
+const staticSegmentValue = 10;
+const splatPenalty = -2;
+const isSplat = (s: string) => s === '*';
+
+function computeScore(path: string, index: boolean | undefined): number {
```

Comment:
It's great that we're now using the same route weight strategy in react-router 6, but my only concern is that we'll have to keep a close eye on if they _change_ that strategy in later versions.

That said, there's not much we can do about it if they're not exporting this scoring method.